File: src/test/java/io/lsdconsulting/lsd/distributed/generator/integration/LsdLoggerIT.java
package io.lsdconsulting.lsd.distributed.generator.integration;

import com.lsd.core.LsdContext;
import com.lsd.core.domain.ComponentName;
import com.lsd.core.domain.SequenceEvent;
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction;
import io.lsdconsulting.lsd.distributed.generator.diagram.InteractionGenerator;
import io.lsdconsulting.lsd.distributed.generator.diagram.LsdLogger;
import io.lsdconsulting.lsd.distributed.generator.integration.testapp.TestApplication;
import io.lsdconsulting.lsd.distributed.generator.integration.testapp.config.RepositoryConfig;
import io.lsdconsulting.lsd.distributed.generator.integration.testapp.repository.TestRepository;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.TestPropertySource;

import java.util.List;
import java.util.Map;
import java.util.Optional;

import static io.lsdconsulting.lsd.distributed.access.model.InteractionType.*;
import static io.lsdconsulting.lsd.distributed.generator.integration.testapp.repository.TestRepository.tearDownDatabase;
import static org.apache.commons.lang3.RandomStringUtils.randomAlphanumeric;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.mockito.Mockito.*;
import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;

@Slf4j
@Import(RepositoryConfig.class)
@SpringBootTest(webEnvironment = RANDOM_PORT, classes = TestApplication.class)
@TestPropertySource("classpath:application-test.properties")
public class LsdLoggerIT {

    @Autowired
    private TestRepository testRepository;

    @Autowired
    private InteractionGenerator interactionGenerator;

    private final String setupTraceId = randomAlphanumeric(10);
    private final String mainTraceId = randomAlphanumeric(10);
    private final String sourceName = randomAlphanumeric(10).toUpperCase();
    private final String targetName = randomAlphanumeric(10).toUpperCase();

    private final LsdContext realContext = new LsdContext();
    private final LsdContext lsdContext = spy(realContext);
    private final ArgumentCaptor<SequenceEvent> eventCaptor = ArgumentCaptor.forClass(SequenceEvent.class);

    private LsdLogger lsdLogger;

    @BeforeEach
    public void setup() {
        lsdLogger = new LsdLogger(interactionGenerator, lsdContext);
    }

    @AfterAll
    static void tearDown() {
        tearDownDatabase();
    }

    @Test
    void shouldLogInteractionsInLsdContextWithSuppliedMultipleTraceIds() {
        testRepository.save(InterceptedInteraction.builder()
                .traceId(mainTraceId)
                .httpMethod("GET")
                .path("/api-listener?message=from_test")
                .serviceName(sourceName)
                .target(targetName)
                .interactionType(REQUEST)
                .build());
        testRepository.save(InterceptedInteraction.builder()
                .traceId(setupTraceId)
                .target("SomethingDoneEvent")
                .serviceName("TestApp")
                .interactionType(PUBLISH)
                .build());
        testRepository.save(InterceptedInteraction.builder()
                .traceId(mainTraceId)
                .serviceName(sourceName)
                .target(targetName)
                .interactionType(RESPONSE)
                .elapsedTime(10L)
                .httpStatus("200 OK")
                .build());
        testRepository.save(InterceptedInteraction.builder()
                .traceId(setupTraceId)
                .serviceName("TestApp")
                .target("SomethingDoneEvent")
                .interactionType(CONSUME)
                .build());
        testRepository.save(InterceptedInteraction.builder()
                .traceId(mainTraceId)
                .httpMethod("POST")
                .path("/external-api?message=from_feign")
                .serviceName("TestApp")
                .target("UNKNOWN_TARGET")
                .interactionType(REQUEST)
                .build());
        testRepository.save(InterceptedInteraction.builder()
                .traceId(mainTraceId)
                .serviceName("TestApp")
                .target("UNKNOWN_TARGET")
                .interactionType(RESPONSE)
                .elapsedTime(20L)
                .httpStatus("200 OK")
                .build());

        lsdLogger.captureInteractionsFromDatabase(Map.of(mainTraceId, Optional.of("blue"), setupTraceId, Optional.of("green")));
        
        verify(lsdContext, times(6)).capture(eventCaptor.capture());

        List<SequenceEvent> capturedValues = eventCaptor.getAllValues();
        assertThat(capturedValues, hasSize(6));
        assertThat(capturedValues.get(0), allOf(
                hasProperty("label", is("GET /api-listener?message=from_test")),
                hasProperty("from", is(new ComponentName(sourceName))),
                hasProperty("to", is(new ComponentName(targetName))),
                hasProperty("colour", is("blue"))
        ));
        assertThat(capturedValues.get(1), allOf(
                hasProperty("label", is("publish event")),
                hasProperty("from", is(new ComponentName("TestApp"))),
                hasProperty("to", is(new ComponentName("SomethingDoneEvent"))),
                hasProperty("colour", is("green"))
        ));
        assertThat(capturedValues.get(2), allOf(
                hasProperty("label", is("sync 200 OK response (10 ms)")),
                hasProperty("from", is(new ComponentName(targetName))),
                hasProperty("to", is(new ComponentName(sourceName))),
                hasProperty("colour", is("blue"))
        ));
        assertThat(capturedValues, hasItem(allOf(
                hasProperty("label", is("consume message")),
                hasProperty("from", is(new ComponentName("SomethingDoneEvent"))),
                hasProperty("to", is(new ComponentName("TestApp"))),
                hasProperty("colour", is("green"))
        )));
        assertThat(capturedValues, hasItem(allOf(
                hasProperty("label", is("POST /external-api?message=from_feign")),
                hasProperty("from", is(new ComponentName("TestApp"))),
                hasProperty("to", is(new ComponentName("UNKNOWN_TARGET"))),
                hasProperty("colour", is("blue"))
        )));
        assertThat(capturedValues, hasItem(allOf(
                hasProperty("label", is("sync 200 OK response (20 ms)")),
                hasProperty("from", is(new ComponentName("UNKNOWN_TARGET"))),
                hasProperty("to", is(new ComponentName("TestApp"))),
                hasProperty("colour", is("blue"))
        )));
    }
}

File: src/test/java/io/lsdconsulting/lsd/distributed/generator/integration/LsdLoggerIT.kt
package io.lsdconsulting.lsd.distributed.generator.integration

import com.lsd.core.LsdContext
import com.lsd.core.domain.ParticipantType.PARTICIPANT
import com.lsd.core.domain.SequenceEvent
import io.lsdconsulting.lsd.distributed.access.model.InteractionType
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction
import io.lsdconsulting.lsd.distributed.generator.diagram.InteractionGenerator
import io.lsdconsulting.lsd.distributed.generator.diagram.LsdLogger
import io.lsdconsulting.lsd.distributed.generator.integration.testapp.TestApplication
import io.lsdconsulting.lsd.distributed.generator.integration.testapp.config.RepositoryConfig
import io.lsdconsulting.lsd.distributed.generator.integration.testapp.repository.TestRepository
import io.mockk.*
import org.apache.commons.lang3.RandomStringUtils
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT
import org.springframework.context.annotation.Import
import org.springframework.test.context.ActiveProfiles
import java.time.ZoneId
import java.time.ZonedDateTime
import java.util.*

@Import(RepositoryConfig::class)
@SpringBootTest(webEnvironment = RANDOM_PORT, classes = [TestApplication::class])
@ActiveProfiles("test")
class LsdLoggerIT {
    @Autowired
    private lateinit var testRepository: TestRepository

    @Autowired
    private lateinit var interactionGenerator: InteractionGenerator

    private val setupTraceId = RandomStringUtils.randomAlphanumeric(10)
    private val mainTraceId = RandomStringUtils.randomAlphanumeric(10)
    private val sourceName = RandomStringUtils.randomAlphanumeric(10).uppercase(Locale.getDefault())
    private val targetName = RandomStringUtils.randomAlphanumeric(10).uppercase(Locale.getDefault())
    private val sequenceEventSlot = ArrayList<SequenceEvent>()
    private val lsdContext = spyk(LsdContext())

    private lateinit var underTest: LsdLogger

    @BeforeEach
    fun setup() {
        underTest = LsdLogger(interactionGenerator, lsdContext)
        every { lsdContext.capture(capture(sequenceEventSlot)) } just Runs
    }

    @Test
    fun `should log interactions in lsd context with supplied multiple trace ids`() {
        val interceptedInteraction1 = InterceptedInteraction(
            traceId = mainTraceId,
            httpMethod = "GET",
            path = "/api-listener?message=from_test",
            serviceName = sourceName,
            target = targetName,
            interactionType = InteractionType.REQUEST,
            elapsedTime = 0,
            createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
        )
        testRepository.save(interceptedInteraction1)
        Thread.sleep(5)

        val interceptedInteraction2 = InterceptedInteraction(
            traceId = setupTraceId,
            target = "SomethingDoneEvent",
            serviceName = "TestApp",
            interactionType = InteractionType.PUBLISH,
            elapsedTime = 0,
            createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
        )
        testRepository.save(interceptedInteraction2)
        Thread.sleep(5)

        val interceptedInteraction3 = InterceptedInteraction(
            traceId = mainTraceId,
            serviceName = sourceName,
            target = targetName,
            interactionType = InteractionType.RESPONSE,
            elapsedTime = 10L,
            httpStatus = "200 OK",
            createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
        )
        testRepository.save(interceptedInteraction3)
        Thread.sleep(5)

        val interceptedInteraction4 = InterceptedInteraction(
            traceId = setupTraceId,
            serviceName = "TestApp",
            target = "SomethingDoneEvent",
            interactionType = InteractionType.CONSUME,
            elapsedTime = 0,
            createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
        )
        testRepository.save(interceptedInteraction4)
        Thread.sleep(5)

        val interceptedInteraction5 = InterceptedInteraction(
            traceId = mainTraceId,
            httpMethod = "POST",
            path = "/external-api?message=from_feign",
            serviceName = "TestApp",
            target = "UNKNOWN_TARGET",
            interactionType = InteractionType.REQUEST,
            elapsedTime = 0,
            createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
        )
        testRepository.save(interceptedInteraction5)
        Thread.sleep(5)

        val interceptedInteraction6 = InterceptedInteraction(
            traceId = mainTraceId,
            serviceName = "TestApp",
            target = "UNKNOWN_TARGET",
            interactionType = InteractionType.RESPONSE,
            elapsedTime = 20L,
            httpStatus = "200 OK",
            createdAt = ZonedDateTime.now(ZoneId.of("UTC"))
        )
        testRepository.save(interceptedInteraction6)
        Thread.sleep(5)

        underTest.captureInteractionsFromDatabase(
            linkedMapOf(
                mainTraceId to Optional.of("blue"),
                setupTraceId to Optional.of("green")
            )
        )

        verify(exactly = 6) { lsdContext.capture(any<SequenceEvent>()) }
        assertThat(
            sequenceEventSlot, contains(
                allOf(
                    hasProperty("label", `is`("GET /api-listener?message=from_test")),
                    hasProperty("from", `is`(PARTICIPANT.called(sourceName))),
                    hasProperty("to", `is`(PARTICIPANT.called(targetName))),
                    hasProperty("colour", `is`("blue"))
                ), allOf(
                    hasProperty("label", `is`("publish event")),
                    hasProperty("from", `is`(PARTICIPANT.called("TestApp"))),
                    hasProperty("to", `is`(PARTICIPANT.called("SomethingDoneEvent"))),
                    hasProperty("colour", `is`("green"))
                ), allOf(
                    hasProperty("label", `is`("sync 200 OK response (10 ms)")),
                    hasProperty("from", `is`(PARTICIPANT.called(targetName))),
                    hasProperty("to", `is`(PARTICIPANT.called(sourceName))),
                    hasProperty("colour", `is`("blue"))
                ),allOf(
                    hasProperty("label", `is`("consume message")),
                    hasProperty("from", `is`(PARTICIPANT.called("SomethingDoneEvent"))),
                    hasProperty("to", `is`(PARTICIPANT.called("TestApp"))),
                    hasProperty("colour", `is`("green"))
                ),allOf(
                    hasProperty("label", `is`("POST /external-api?message=from_feign")),
                    hasProperty("from", `is`(PARTICIPANT.called("TestApp"))),
                    hasProperty("to", `is`(PARTICIPANT.called("UNKNOWN_TARGET"))),
                    hasProperty("colour", `is`("blue"))
                ),allOf(
                    hasProperty("label", `is`("sync 200 OK response (20 ms)")),
                    hasProperty("from", `is`(PARTICIPANT.called("UNKNOWN_TARGET"))),
                    hasProperty("to", `is`(PARTICIPANT.called("TestApp"))),
                    hasProperty("colour", `is`("blue"))
                )
            )
        )
    }

    companion object {
        @JvmStatic
        @AfterAll
        fun tearDown() {
            TestRepository.tearDownDatabase()
        }
    }
}
